Batteries/Data/Fin/Lemmas.lean

```diff
@@ -42,63 +42,70 @@ theorem findSome?_succ_of_none {f : Fin (n+1) → Option α} (h : f 0 = none) :
 theorem findSome?_succ_of_isNone {f : Fin (n+1) → Option α} (h : (f 0).isNone) :
     findSome? f = findSome? fun i => f i.succ := by simp_all [findSome?_succ_of_none]
 
-theorem exists_of_findSome?_eq_some {f : Fin n → Option α} (h : findSome? f = some x) :
-    ∃ i, f i = some x := by
+@[simp]
+theorem findSome?_eq_some_iff {f : Fin n → Option α} :
+    findSome? f = some a ↔ ∃ i, f i = some a ∧ ∀ j < i, f j = none := by
   induction n with
-  | zero => rw [findSome?_zero] at h; contradiction
+  | zero =>
+    simp only [findSome?_zero, (Option.some_ne_none _).symm, false_iff]
+    exact fun  ⟨i, _⟩ => i.elim0
   | succ n ih =>
-    rw [findSome?_succ] at h
-    match heq : f 0 with
-    | some x =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_some] at h
-      exists 0
-      rw [heq, h]
-    | none =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_none] at h
-      match ih h with | ⟨i, _⟩ => exists i.succ
+    simp only [findSome?_succ, Option.orElse_eq_orElse, Option.orElse_eq_or, Option.or_eq_some_iff,
+      ih, forall_fin_succ, exists_fin_succ, false_implies, not_lt_zero, implies_true,
+      and_true, succ_pos, forall_const, succ_lt_succ_iff, and_left_comm (b := f 0 = none),
+      exists_and_left]
 
-theorem eq_none_of_findSome?_eq_none {f : Fin n → Option α} (h : findSome? f = none) (i) :
-    f i = none := by
-  induction n with
-  | zero => cases i; contradiction
-  | succ n ih =>
-    rw [findSome?_succ] at h
-    match heq : f 0 with
-    | some x =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_some] at h
-      contradiction
-    | none =>
-      rw [heq, Option.orElse_eq_orElse, Option.orElse_none] at h
-      cases i using Fin.cases with
-      | zero => exact heq
-      | succ i => exact ih h i
-
-@[simp] theorem findSome?_isSome_iff {f : Fin n → Option α} :
-    (findSome? f).isSome ↔ ∃ i, (f i).isSome := by
-  simp only [Option.isSome_iff_exists]
-  constructor
-  · intro ⟨x, hx⟩
-    match exists_of_findSome?_eq_some hx with
-    | ⟨i, hi⟩ => exists i, x
-  · intro ⟨i, x, hix⟩
-    match h : findSome? f with
-    | some x => exists x
-    | none => rw [eq_none_of_findSome?_eq_none h i] at hix; contradiction
+@[simp]
+theorem findSome?_isSome_iff {f : Fin n → Option α} :
+    (findSome? f).isSome ↔ ∃ i, (f i).isSome ∧ ∀ j < i, (f j).isNone := by
```

Comment:
This variant is not stronger than `(findSome? f).isSome ↔ ∃ i, (f i).isSome`: if there is such an `i` then of course there must be a first one.

Reply:
I agree, but I want to avoid having to re-derive this fact every time.

Reply:
Use `findSome?_eq_some_iff` instead? Also note that the <- direction is strictly weaker than the <- direction of `(findSome? f).isSome ↔ ∃ i, (f i).isSome`.